test/Microsoft.AspNet.Mvc.Core.Test/Filters/FormatFilterTest.cs

```diff
@@ -0,0 +1,324 @@
+        [InlineData("json", FormatSource.QueryData, "application/json")]
+        public void FormatFilter_ContextContainsFormat_ContainsProducesFilter_Matching(
+            string format, 
+            FormatSource place, 
+            string contentType)
+        {
+            // Arrange
+            var produces = new ProducesAttribute(contentType, new string[] { "application/foo", "text/bar" });
+            var context = CreateResourceExecutingContext(new IFilter[] { produces }, format, place);
+            var filter = new FormatFilterAttribute();
+
+            // Act
+            filter.OnResourceExecuting(context);
+
+            // Assert
+            var result = Assert.IsType<HttpNotFoundResult>(context.Result);
+        }
+
+        [Theory]
+        [InlineData("json", FormatSource.RouteData, "application/json")]
+        [InlineData("json", FormatSource.QueryData, "application/json")]
+        public void FormatFilter_ContextContainsFormat_ContainsProducesFilter_Conflicting(
+            string format,
+            FormatSource place,
+            string contentType)
+        {
+            // Arrange
+            var mediaType = MediaTypeHeaderValue.Parse(contentType);
+            var produces = new ProducesAttribute("application/xml", new string[] { "application/foo", "text/bar" });
+            var context = CreateResourceExecutingContext(new IFilter[] { produces }, format, place);
+            var filter = new FormatFilterAttribute();
+
+            // Act
+            filter.OnResourceExecuting(context);
+
+            // Assert
+            var result = Assert.IsType<HttpNotFoundResult>(context.Result);
+        }
+
+        [Theory]
+        [InlineData("", FormatSource.RouteData)]
+        [InlineData(null, FormatSource.QueryData)]
+        [InlineData("", FormatSource.RouteData)]
+        [InlineData(null, FormatSource.QueryData)]
+        public void FormatFilter_ContextContainsFormat_Invalid(
+            string format,
+            FormatSource place)
+        {
+            // Arrange            
+            var resourceExecutingContext = CreateResourceExecutingContext(
+                new IFilter[] { }, 
+                format, 
+                FormatSource.RouteData);
+            var filter = new FormatFilterAttribute();
+
+            // Act
+            filter.OnResourceExecuting(resourceExecutingContext);
+
+            var result = Assert.IsType<HttpNotFoundResult>(resourceExecutingContext.Result);
+        }
+
+        private static ResourceExecutingContext CreateResourceExecutingContext(
+            IFilter[] filters,
+            string format = null, 
+            FormatSource? place = null)
+        {
+            if(format == null || place == null)
+            {
+                var context = new ResourceExecutingContext(
+                    CreateActionContext(),
+                    filters);
+                context.Result = new HttpStatusCodeResult(200);
```

Comment:
Don't set `Result`. Setting `Result` is a short-circuit of the filter pipeline. You'll never see this in an actual request.